### ISSUE TITLE
yoloposeV8X training was too slow

### ISSUE THREAD

**Issue by ZhangLe-fighting** [ultralytics/ultralytics#3863]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

The GPU utilization rate is too low during the training process, and the training is too slow，May I ask what the reason is！

Training Command：
from ultralytics import YOLO
YOLO(model).train(**args)

yolopose.yaml
# Ultralytics YOLO , AGPL-3.0 license
# Default training settings and hyperparameters for medium-augmentation COCO training

task: pose  # (str) YOLO task, i.e. detect, segment, classify, pose
mode: train  # (str) YOLO mode, i.e. train, val, predict, export, track, benchmark

# Train settings -------------------------------------------------------------------------------------------------------
model: yolov8x-pose.pt # (str, optional) path to model file, i.e. yolov8n.pt, yolov8n.yaml
data: coco-pose.yaml # (str, optional) path to data file, i.e. coco128.yaml
epochs: 350  # (int) number of epochs to train for
patience: 150  # (int) epochs to wait for no observable improvement for early stopping of training
batch: 64 # (int) number of images per batch (-1 for AutoBatch)
imgsz: 640  # (int | list) input images size as int for train and val modes, or list[w,h] for predict and export modes
save: True  # (bool) save train checkpoints and predict results
save_period: 10 # (int) Save checkpoint every x epochs (disabled if < 1)
cache: True  # (bool) True/ram, disk or False. Use cache for data loading
device: 0,1,2,3,4,5,6,7 # (int | str | list, optional) device to run on, i.e. cuda device=0 or device=0,1,2,3 or device=cpu
workers: 8  # (int) number of worker threads for data loading (per RANK if DDP)
project: yolov8x-pose # (str, optional) project name
name: coco-pose # (str, optional) experiment name, results saved to 'project/name' directory
exist_ok: False  # (bool) whether to overwrite existing experiment
pretrained: True  # (bool | str) whether to use a pretrained model (bool) or a model to load weights from (str)
optimizer: NAdam  # (str) optimizer to use, choices=[SGD, Adam, Adamax, AdamW, NAdam, RAdam, RMSProp, auto]
verbose: True  # (bool) whether to print verbose output
seed: 0  # (int) random seed for reproducibility
deterministic: False  # (bool) whether to enable deterministic mode
single_cls: False  # (bool) train multi-class data as single-class
rect: False  # (bool) rectangular training if mode='train' or rectangular validation if mode='val'
cos_lr: True  # (bool) use cosine learning rate scheduler
close_mosaic: 10  # (int) disable mosaic augmentation for final epochs
resume: True  # (bool) resume training from last checkpoint
amp: False  # (bool) Automatic Mixed Precision (AMP) training, choices=[True, False], True runs AMP check
fraction: 1.0  # (float) dataset fraction to train on (default is 1.0, all images in train set)
profile: False  # (bool) profile ONNX and TensorRT speeds during training for loggers
# Segmentation
overlap_mask: True  # (bool) masks should overlap during training (segment train only)
mask_ratio: 4  # (int) mask downsample ratio (segment train only)
# Classification
dropout: 0.0  # (float) use dropout regularization (classify train only)

# Val/Test settings ----------------------------------------------------------------------------------------------------
val: True  # (bool) validate/test during training
split: val  # (str) dataset split to use for validation, i.e. 'val', 'test' or 'train'
save_json: False  # (bool) save results to JSON file
save_hybrid: False  # (bool) save hybrid version of labels (labels + additional predictions)
conf:  # (float, optional) object confidence threshold for detection (default 0.25 predict, 0.001 val)
iou: 0.7  # (float) intersection over union (IoU) threshold for NMS
max_det: 300  # (int) maximum number of detections per image
half: False  # (bool) use half precision (FP16)
dnn: False  # (bool) use OpenCV DNN for ONNX inference
plots: True  # (bool) save plots during train/val

# Prediction settings --------------------------------------------------------------------------------------------------
source:  # (str, optional) source directory for images or videos
show: False  # (bool) show results if possible
save_txt: False  # (bool) save results as .txt file
save_conf: False  # (bool) save results with confidence scores
save_crop: False  # (bool) save cropped images with results
show_labels: True  # (bool) show object labels in plots
show_conf: True  # (bool) show object confidence scores in plots
vid_stride: 1  # (int) video frame-rate stride
line_width:   # (int, optional) line width of the bounding boxes, auto if missing
visualize: False  # (bool) visualize model features
augment: False  # (bool) apply image augmentation to prediction sources
agnostic_nms: False  # (bool) class-agnostic NMS
classes:  # (int | list[int], optional) filter results by class, i.e. class=0, or class=[0,2,3]
retina_masks: False  # (bool) use high-resolution segmentation masks
boxes: True  # (bool) Show boxes in segmentation predictions

# Export settings ------------------------------------------------------------------------------------------------------
format: torchscript  # (str) format to export to, choices at https://docs.ultralytics.com/modes/export/#export-formats
keras: False  # (bool) use Kera=s
optimize: False  # (bool) TorchScript: optimize for mobile
int8: False  # (bool) CoreML/TF INT8 quantization
dynamic: False  # (bool) ONNX/TF/TensorRT: dynamic axes
simplify: False  # (bool) ONNX: simplify model
opset:  # (int, optional) ONNX: opset version
workspace: 4  # (int) TensorRT: workspace size (GB)
nms: False  # (bool) CoreML: add NMS

# Hyperparameters ------------------------------------------------------------------------------------------------------
lr0: 0.001  # (float) initial learning rate (i.e. SGD=1E-2, Adam=1E-3)
lrf: 0.005  # (float) final learning rate (lr0 * lrf)
momentum: 0.920  # (float) SGD momentum/Adam beta1
weight_decay: 0.0003  # (float) optimizer weight decay 5e-4
warmup_epochs: 5.0  # (float) warmup epochs (fractions ok)
warmup_momentum: 0.6  # (float) warmup initial momentum
warmup_bias_lr: 0.05  # (float) warmup initial bias lr
box: 7.5  # (float) box loss gain
cls: 0.5  # (float) cls loss gain (scale with pixels)
dfl: 1.5  # (float) dfl loss gain
pose: 12.0  # (float) pose loss gain
kobj: 1.0  # (float) keypoint obj loss gain
label_smoothing: 0.001  # (float) label smoothing (fraction)
nbs: 64  # (int) nominal batch size
hsv_h: 0.015  # (float) image HSV-Hue augmentation (fraction)
hsv_s: 0.7  # (float) image HSV-Saturation augmentation (fraction)
hsv_v: 0.4  # (float) image HSV-Value augmentation (fraction)
degrees: 30  # (float) image rotation (+/- deg)
translate: 0.1  # (float) image translation (+/- fraction)
scale: 0.5  # (float) image scale (+/- gain)
shear: 10  # (float) image shear (+/- deg)
perspective: 0.0002  # (float) image perspective (+/- fraction), range 0-0.001
flipud: 0.01  # (float) image flip up-down (probability)
fliplr: 0.5  # (float) image flip left-right (probability)
mosaic: 1.0  # (float) image mosaic (probability)
mixup: 0.001  # (float) image mixup (probability)
copy_paste: 0.01  # (float) segment copy-paste (probability)

# Custom config.yaml ---------------------------------------------------------------------------------------------------
cfg:  # (str, optional) for overriding defaults.yaml

# Debug, do not modify -------------------------------------------------------------------------------------------------
v5loader: False  # (bool) use legacy YOLOv5 dataloader (deprecated)

# Tracker settings ------------------------------------------------------------------------------------------------------
tracker: botsort.yaml  # (str) tracker type, choices=[botsort.yaml, bytetrack.yaml]

![2](https://github.com/ultralytics/ultralytics/assets/55420158/78471cf5-01c1-4ef0-8fd3-ff12cf1fe7c5)

**Comment by glenn-jocher** [comment]
@ZhangLe-fighting hello, thank you for raising this issue.

The slower training process and low GPU utilization rate may be due to a variety of reasons. Let's explore a few situations that might be causing this:

1. **Data Loading Speed**: Check if your data loading speed is fast enough. If it's too slow, it might be causing a bottleneck. Increasing the number of worker threads may potentially resolve this problem.

2. **Hardware Architecture Compatibility**: Not all algorithms are optimized for all GPU architectures. YOLOv8 might not be fully optimized for your specific GPU architecture which could be causing the low utilization rate.

3. **Batch Size**: The batch size is another factor that can affect the GPU utilization. Try varying your batch size to see if that improves the GPU utilization.

4. **Training Configuration**: Some training configurations (such as optimizer choice, learning rate, etc.) might influence the speed of the training process. 

5. **GPU Device Usage**: Make sure no other heavy processes are running on the same GPU devices that can affect overall GPU utilization.

6. **Software Settings**: Check whether you configured your software environment correctly. Make sure your CUDA and cuDNN versions are compatible with the PyTorch version you're using.

Exploring these areas should help you find the source of the problem. I hope this helps and that you find a solution to your problem soon! Let us know if you need further assistance.

**Comment by ZhangLe-fighting** [comment]
Thank you for your reply. I will conduct testing based on your suggestion. In addition, I would like to inquire about a question: is the yolov8x pose model provided in readme a trained model based on the coco keypoint dataset? I used this model as a pre training model and continued to use the coco keypoint dataset for training. I found that the size of the model (796MB) is much larger than the model you provided (132MB). This seems to be a bit problematic. Could you please help me answer this question!

**Comment by glenn-jocher** [comment]
Hello @ZhangLe-fighting, I'm glad you found the previous information helpful and plan to conduct testing.

To answer your new question, the YOLOv8x pose model mentioned in the readme likely utilized the COCO Keypoint dataset for its training. Pre-trained models in general are trained on a large dataset followed by saving the weights for further fine-tuning or direct usage.

The increased model size you observed could be due to reasons such as:

1. **Continued Training**: When training continues on an existing model, new weights are constantly updated, possibly leading to an increase in file size due to additional metadata.

2. **Different Configuration Settings**: Various aspects like the number of layers, neurons per layer or finer settings like precision (float16, float32 etc) can influence the size of the model.

For your concern, please also check if there was saving of unnecessary information (like gradients) during your training process, which might increase the size of the saved model.

I hope this helps your understanding and your project. Do reach out for further clarification, if required.

**Comment by ZhangLe-fighting** [comment]
Thank you very much for your response. It has been very helpful to me.

**Comment by glenn-jocher** [comment]
Hello @ZhangLe-fighting, 

I'm really pleased to hear that you found the information helpful! If you have any other questions or run into any further issues, don't hesitate to ask. Remember, there's no such thing as a silly question, and we're here to support you. Happy coding and best of luck with your project!
